Transfer CalcJob

ramirezfranciscof opened the pull request:
Fixes #4600 

This is a utility that provides users a way to locally retrieve files that were left in remote folders after calculations and put them back in the remote machine once they are needed.

You can see an explanation of its interface and how it works in [this document](https://docs.google.com/document/d/1b9pqTfwo5yxegQ72hGXQ9OBQNPhfOZIXP3XTD-_n7hY/edit). You can leave there your comments on the interface. In the meantime I will think of examples that don't include QE to put in the AiiDA doc as well as tests for this, but before incorporating those here I wanted to make sure we all agreed on the interface first.

You can also try it yourselves by running the workflow example found [in the same folder](https://drive.google.com/drive/u/1/folders/1vBxwjnJmACydkbm4PuVYubT8gAwh4il8), that illustrates how to locally save the necessary restart files for QE and set them again in the remote to lunch a restart.

sphuber commented:
@ramirezfranciscof could you please already rebase this on the latest `develop`, squashing the commits and writing a descriptive commit message, and make sure that the tests run.

ramirezfranciscof commented:
Ah, sure, I didn't notice this. I'll rebase, thanks for the heads up.

Regarding the squashing: do I need to do this now? This should only matter when we are merging, and since there might still be some corrections to make does it make any difference? (as a matter of fact I'm not even sure that we want to keep any of this commits or just squash everything; or at least everything so far, maybe test and doc could be separate)

sphuber commented:
> Regarding the squashing: do I need to do this now? This should only matter when we are merging, and since there might still be some corrections to make does it make any difference? (as a matter of fact I'm not even sure that we want to keep any of this commits or just squash everything; or at least everything so far, maybe test and doc could be separate)

Sure, there may be additional changes that need to be done during the review, but squashing now and writing the corresponding commit messages, forces you to think how to describe the changes and why they are important. Of course they might be explained in detail in the docs and docstrings, but it is important that the commit message also contains a resume of why this feature is being added and describe the main design decisions made. This not only makes it easy for the reviewer to get a quick context of what is happening but already prepares the text for when it is time to merge.

ramirezfranciscof commented:
Ok, now everything works locally and the docstring should be fixed. I also squashed into two distinct commits. Anyways this push was mainly for people to be able to access the code if they want to try to run the example workflow, but I think the idea for now was that they provide the feedback mostrly on the interface explained in the [linked document](https://docs.google.com/document/d/1b9pqTfwo5yxegQ72hGXQ9OBQNPhfOZIXP3XTD-_n7hY/edit).

giovannipizzi commented:
@ramirezfranciscof is there still work to be done with this PR before making a proper PR from a Draft one?

ramirezfranciscof commented:
I think I just need to add the tests. When @mkotiuga tried it she seemed to have problems when deleting nodes only when using this branch, but I've been trying to reproduce the error and just can't find anything, so I might check with her one more time.

ramirezfranciscof commented:
So, I just want to keep a record that I tried to use `tmp_path` as explained [here](https://stackoverflow.com/a/36071167/638366) but I keep getting an error that saying something like the pathlib object does not have a join property, so I ended up using something like the examples [here](https://docs.pytest.org/en/stable/tmpdir.html). I couldn't find any better source on how to use this fixture.

sphuber commented:
Not sure what you mean, the tests you have added look fine. We're you perhaps trying to do `path = tmp_path.join('other_path')`? Anyway, the way you use it `path = tmp_path / 'other_path'` is exactly how it should be used.

ramirezfranciscof commented:
@sphuber 

I have a question for this. For how `task_update_job` seems to work, I understand that if you wrap a variable around `raise Return( ... )`, this is returned from the coroutine as if it was a normal return statement. Indeed here there seem to be two passes, one from `do_update` to `task_update_job`, and one from `task_update_job` to the outside world.

```
def task_update_job(...):
    ...
    def do_update():
        ...
        raise Return(job_done)
    ...
    job_done = yield exponential_backoff_retry(do_update, ...)
    ...
    raise Return(job_done)

...
# Externally:
job_done = yield self._launch_task(task_update_job, node, self.process.runner.job_manager)
...
```

The coroutine `task_retrieve_job` however, does something differently. The internal `do_retrieve` returns a value but this is NOT catched by the `task_retrieve_job`, nor is anything caught outside of this last one:

```
def task_retrieve_job(...):
    ...
    def do_retrieve():
        ...
        raise Return(execmanager.retrieve_calculation(...))
    ...
    yield exponential_backoff_retry(do_retrieve, ...)
    ...
    raise Return
...
# Externally:
yield self._launch_task(task_retrieve_job, node, transport_queue, temp_folder)
...
```

Finally, in `task_upload_job` the opposite happens: `do_upload` does not seem to return anything but `task_upload_job` is catching something to return (although this is not being caught outside either).

```
def task_upload_job(...):
    ...
    def do_upload():
        ...
        raise Return
    ...
    result = yield exponential_backoff_retry(do_upload, ...)
    ...
    raise Return(result)
...
# Externally:
yield self._launch_task(task_upload_job, self.process, transport_queue)
...
```

Is this last `result` variable fulfiling any purpose? Is this related to how the `raise Result(...)` mechanism work that is important? Can I just replace this result caught from `do_upload` and returned by `task_upload_job` by another variable that I actually need to determine inside the former but use outside the latter?

Since I ommited a lot of content (which I thought was not relevant but I might be wrong), here are the sections of the code referenced:

* `task_upload_job` goes from [line 45 to line 103](https://github.com/aiidateam/aiida-core/blob/v1.5.0/aiida/engine/processes/calcjobs/tasks.py#L45-L103) of `engine/processes/calcjobs/tasks.py`.
* `task_update_job` goes from [line 154 to line 213](https://github.com/aiidateam/aiida-core/blob/v1.5.0/aiida/engine/processes/calcjobs/tasks.py#L154-L213) of `engine/processes/calcjobs/tasks.py`.
* `task_retrieve_job` goes from [line 216 to line 275](https://github.com/aiidateam/aiida-core/blob/v1.5.0/aiida/engine/processes/calcjobs/tasks.py#L216-L275) of `engine/processes/calcjobs/tasks.py`.
* The external call to all of these goes from [line 45 to line 103](https://github.com/aiidateam/aiida-core/blob/v1.5.0/aiida/engine/processes/calcjobs/tasks.py#L45-L103) of `engine/processes/calcjobs/tasks.py`.

ramirezfranciscof commented:
@sphuber I just implemented the skip of the submit/update steps and added the integration test we discussed. Probably this is not the best place for it, but I didn't know exactly where to put it, so I'll wait for your feedback on this.

sphuber commented on aiida/common/datastructures.py:
```diff
@@ -98,7 +98,8 @@ class CalcInfo(DefaultFieldsAttributeDict):
         'codes_info',
-        'codes_run_mode'
+        'codes_run_mode',
+        'skip_submit'
```
Please add a description of this new field in the docstring.

sphuber commented on aiida/calculations/transfer.py:
```diff
@@ -0,0 +1,250 @@
+from aiida.common.datastructures import CalcInfo
+
+
+def validate_instructions(instructions, ctx):  # pylint: disable=inconsistent-return-statements, unused-argument
```
```suggestion
def validate_instructions(instructions, _):
```

ramirezfranciscof replied:
Hahah, NOW I see what you meant in the [other PR](https://github.com/aiidateam/aiida-core/pull/4437#discussion_r526668141) 👍

sphuber commented on aiida/calculations/transfer.py:
```diff
@@ -0,0 +1,250 @@
+
+    if not any([local_files, remote_files, symlink_files]):
+        errmsg = (
+            'no indication of which files to copy were found in the instruction node {}.\n'
```
```suggestion
            f'no indication of which files to copy was found in the instructions node: {instructions}.\n'
```

sphuber commented on aiida/calculations/transfer.py:
```diff
@@ -0,0 +1,250 @@
+    if retrieve_files is None:
+        errmsg = (
+            '\n\n'
+            'no indication of what to do in the instruction node:\n > {}\n'
```
```suggestion
            f'no indication of what to do in the instruction node: {instructions}\n'
```

sphuber commented on aiida/calculations/transfer.py:
```diff
@@ -0,0 +1,250 @@
+            'to copy them to the specified folder on the remote computer,\n'
+            'set it to False)\n'
+        )
+        return errmsg.format(instructions.uuid)
```
```suggestion
        return errmsg
```

sphuber commented on aiida/calculations/transfer.py:
```diff
@@ -0,0 +1,250 @@
+
+    if not isinstance(retrieve_files, bool):
+        errmsg = (
+            'entry for retrieve files inside of instruction node {} must be\n'
+            'either True or False; instead, it is: {}'
+        )
+        return errmsg.format(instructions.uuid, retrieve_files)
```
```suggestion
            f'entry for retrieve files inside of instruction node {instructions} must be\n'
            f'either True or False; instead, it is: {retrieve_files}'
        )
        return errmsg
```

sphuber commented on aiida/calculations/transfer.py:
```diff
@@ -0,0 +1,250 @@
+            'These should be lists containing tuples following the pattern:\n'
+            '[ ... (source_node_key, source_relpath, target_relpath) ... ] \n'
+        )
+        return errmsg.format(instructions.uuid)
```
```suggestion
        return errmsg
```

sphuber commented on aiida/calculations/transfer.py:
```diff
@@ -0,0 +1,250 @@
+        errmsg = (
+            'no indication of which files to copy were found in the instruction node {}.\n'
+            'Please include at least one of `local_files`, `remote_files`, or `symlink_files`.\n'
+            'These should be lists containing tuples following the pattern:\n'
+            '[ ... (source_node_key, source_relpath, target_relpath) ... ] \n'
```
```suggestion
            'These should be lists containing tuples with format: (source_node_key, source_relpath, target_relpath)\n'
```

sphuber commented on aiida/calculations/transfer.py:
```diff
@@ -0,0 +1,250 @@
+
+    .. code-block:: python
+
+        [ ..., ( source_node_key, path_to_file_in_source, path_to_file_in_target), ... ]
```
```suggestion
        ( source_node_key, path_to_file_in_source, path_to_file_in_target)
```

sphuber commented on aiida/calculations/transfer.py:
```diff
@@ -0,0 +1,250 @@
+        )
+
+        # The transfer just needs a computer, the code are resources are set here
+        spec.inputs['code'].required = False
```
if we are not going to use a code, maybe it is just better to get rid of the port altogether, i.e. `spec.inputs.pop('code', None)`

ramirezfranciscof replied:
I mean, since the original idea involved leaving this open for including in the future the posibility of performing the copy via a script (for cases like CSCS different storages, tapes, etc) then I'm not sure it makes sense to take it out if it will be eventualy be put back in (although now  I can't see specific cases, I guess it technically increases the chances of this later becoming a backwards incompatible change).

sphuber replied:
I would see it from another point of view: if it is not used now whatsoever, and _if_ a code is specified, it will simply be ignored, I don't see the need to have it there, it can just create confusion. On top of that, if in the future we will add support, it can be added without breaking backwards compatibility. So I would really remove it

sphuber commented on aiida/calculations/transfer.py:
```diff
@@ -0,0 +1,250 @@
+        return errmsg.format(instructions.uuid)
+
+
+def validate_transfer_inputs(inputs, ctx):  # pylint: disable=inconsistent-return-statements, unused-argument
```
```suggestion
def validate_transfer_inputs(inputs, _):
```

sphuber commented on aiida/calculations/transfer.py:
```diff
@@ -0,0 +1,250 @@
+
+
+def validate_transfer_inputs(inputs, ctx):  # pylint: disable=inconsistent-return-statements, unused-argument
+    """Check that the instructions dict and the source nodes are consistent"""
```
Also use f-strings as much as possible in this method

sphuber commented on aiida/calculations/transfer.py:
```diff
@@ -0,0 +1,250 @@
+    computer = inputs['metadata']['computer']
+
+    instructions_dict = instructions.get_dict()
+    local_files = instructions_dict.get('local_files', list())
+    remote_files = instructions_dict.get('remote_files', list())
+    symlink_files = instructions_dict.get('symlink_files', list())
```
```suggestion
    local_files = instructions_dict.get('local_files', [])
    remote_files = instructions_dict.get('remote_files', [])
    symlink_files = instructions_dict.get('symlink_files', [])
```

sphuber commented on aiida/calculations/transfer.py:
```diff
@@ -0,0 +1,250 @@
+    new remote folder saved in a RemoteData node).
+
+    Only files from the local computer and a single remote computer can be moved with a single
+    instance of this CalcJob (so you can't directly move files from remote to remote and all
```
This isn't fully clear. When you say "you can't directly move files from remote to remote" that is not really true, because you can move files from a remote folder to another remote folder, right? Only restriction is just that you cannot transfer between different remote computers. So I would simplify this entire paragraph, you can move files between remote folders as long as they are on the same computer.

sphuber commented on aiida/calculations/transfer.py:
```diff
@@ -0,0 +1,250 @@
+    new folder in the remote machine (RemoteData) and put all the files there and will either:
+
+        (1) leave them there (``retrieve_files = False``) or ...
+        (2) create a local FolderData and copy there all the files (``retrieve_files = True``)
```
```suggestion
        (2) retrieve all the files and store them locally in a ``FolderData`` (``retrieve_files = True``)
```

sphuber commented on aiida/calculations/transfer.py:
```diff
@@ -0,0 +1,250 @@
+
+        calc_info = CalcInfo()
+        calc_info.skip_submit = True
+        calc_info.codes_info = []
```
does this need to be set? Or can we omit this line?

ramirezfranciscof replied:
You mean `calc_info.codes_info = []`? I get an error when I remove it (it is a bit long log so I'm not sure it makes sense to copy here; I can't really identify the specific location of the error from the log).

sphuber replied:
Same here, instead of starting to refactor the `presubmit` let's leave this for now

sphuber commented on tests/calculations/test_transfer.py:
```diff
@@ -0,0 +1,152 @@
+    from aiida.calculations.transfer import TransferCalculation
+    from aiida.engine import run
+
+    content_local0 = 'Content of local file'
```
why the `0` suffix? I take it because you want `local` to be the same length as `remote`. I would not do this personally as I think it is confusing. I was looking for the other "local" node and was confused when I didn't find one

sphuber commented on tests/calculations/test_transfer.py:
```diff
@@ -0,0 +1,152 @@
+    content_remote = 'Content of remote file'
+    srcfile_remote = tmp_path / 'file_remote.txt'
+    srcfile_remote.write_text(content_remote)
+    srcnode_remote = orm.RemoteData(computer=aiida_localhost, remote_path=str(tmp_path))
```
Note that we are testing the mechanism here where both the local and remote machine are the localhost. I am thinking whether this is ok in terms of whether there is now code that is untested when there are two different computers. I think having just this single test case should be fine, but it would be easy to parametrize this test over two different computers: one that is localhost, and a different computer (still localhost but different `Computer` instance) but then with `SshTransport`. You could update the `aiida_localhost` fixture to make the transport configurable

ramirezfranciscof replied:
>  I am thinking whether this is ok in terms of whether there is now code that is untested when there are two different computers.

I'm not sure what you mean by this; other than when the transfer is checking that all remote folders belong to the same computer, when does the code care that the `localhost` computer is the same or not as the one where the aiida database is deployed? The engine will just treat all target machines the same, it does not make a difference wether the machine is actually the localhost, no?

I mean, I could add this test, I just don't really understand what difference we are testing for (other than perhaps an integrated test between the engine and the ssh transport).

sphuber replied:
I am not sure either, as I said in my comment I was thinking whether there could be some code that is dependent on the transport being used, that is not itself transport code, that could not be tested now. But since I cannot think of it now and you can't either, fine to leave as is

sphuber commented on aiida/calculations/transfer.py:
```diff
@@ -0,0 +1,250 @@
+        # The transfer just needs a computer, the code are resources are set here
+        spec.inputs['code'].required = False
+        spec.inputs['metadata']['computer'].required = True
+        spec.inputs['metadata']['options']['resources'].default = {
```
Same here, since we are not going to use `resources` can we just pop this port so we also don't have to specify the default?

ramirezfranciscof replied:
I guess the same argument as [the one for code](https://github.com/aiidateam/aiida-core/pull/4194#discussion_r527602924) would apply here. Although now that we are skipping the submit/update, I could instead of setting a default do `required = False`.

ramirezfranciscof replied:
Mmm, so apparently the `scheduler.preprocess_resources` is being called by the `presubmit` routine, so it is still raising errors here...I guess I could try to disable this when `skip_submit` is true, let me try.

ramirezfranciscof replied:
So, from what I could see, the problem is that the presubmit step does a lot of scheduler setups which are intercalated with other things like copy_list checks. It is hard to say on a first approximation which ones can be taken out by an if clause without impacting on the rest of the code. In principle, comenting [these lines](https://github.com/aiidateam/aiida-core/blob/f945dcabbf82a921e9b844b1865681e1a945ef67/aiida/engine/processes/calcjobs/calcjob.py#L506-L516) seems to work:

```python
        if not calc_info.skip_submit:
            resources = self.node.get_option('resources')
            scheduler.preprocess_resources(resources, computer.get_default_mpiprocs_per_machine())
            job_tmpl.job_resource = scheduler.create_job_resource(**resources)

            subst_dict = {'tot_num_mpiprocs': job_tmpl.job_resource.get_tot_num_mpiprocs()}

            for key, value in job_tmpl.job_resource.items():
                subst_dict[key] = value
            mpi_args = [arg.format(**subst_dict) for arg in computer.get_mpirun_command()]
            extra_mpirun_params = self.node.get_option('mpirun_extra_params')  # same for all codes in the same calc
```

However, I feel like just if-ing like this would unnecessarily add complexity "gunk" to the code and would just complicate further modifications of the section. On the other hand, doing a full reorganization of this routine would involve a substantial ammount of work which I'm not sure is fully related to the objective of this PR. I can try to do this if you consider this critical.

sphuber replied:
Fair point, I agree then that this is not the place to do it, so you can consider this comment resolved and leave the code here as is

sphuber reviewed (changes requested):
Thanks @ramirezfranciscof . Implementation of `skip_submit` looks good. Just some minor suggestions. Mostly using f-strings now that we support that and maybe extend the integration test to also include one with a different remote computer.

ramirezfranciscof commented:
@sphuber friendly reminder of this good-ol' PR.

sphuber commented:
@ramirezfranciscof apologies for the delay. I responded to your comments and resolved all outstanding conversations except for one. If you can update the branch and address that final one, this would be good for me to merge. I will just quickly ping @giovannipizzi if he still wanted to check, but if not, I will merge this. Thanks

giovannipizzi commented:
Sorry, I won't have time now - in order not to further delay, maybe just make sure it's marked as beta so we can change the API later? (e.g. to extend to stashing etc.)?

sphuber commented:
@ramirezfranciscof I cannot approve or merge this until you have rebased and addressed the conflicts.

ramirezfranciscof commented:
@sphuber There is one test still running but should pass. I just rebased and adapted the modifications to the changes in the engine, but although you already saw it maybe you could give that a second quick pass in case you notice something of these adaptations. Other than that, the only new thing is the doc section in the howto (work with) data.

sphuber commented on docs/source/howto/data.rst:
```diff
@@ -805,3 +805,66 @@ This command will delete both the file repository and the database.
+    In [2]: transfer_builder = CalculationFactory('core.transfer').get_builder()
+        ... transfer_builder.instructions = instructions_node
+        ... transfer_builder.source_nodes = {'node_keyname': source_node}
+        ... transfer_builder.metadata.computer = computer_node
```
a `Computer` instance is not really a node. It doesn't live in the node table and also is not a subclass of `Node`. You also don't explain how to get it.
```suggestion
        ... transfer_builder.metadata.computer = source_node.computer
```

ramirezfranciscof replied:
So, the thing is that this fix works for when the `source_node` is of type `RemoteData` (which is the case for the first example). However, in the second example I say they can do the same thing but just changing the `dict` setup, so if they copy this for the second part it won't work. There they have to load the remote computer they want to use and pre-set it themselves, I cannot give further details of which one they want to use. I can remove the `_node` part though.

sphuber replied:
I would actually keep the change I suggested and then simply explain what you just did in the example with the `FolderData` source node. There I would then suggest to show them to use `load_computer('some-computer-label')`. This is the typical way to load a `Computer` instance through the API. You can make it as explicit as the following for clarity
```
transfer_builder.metadata.computer = load_computer('some-computer-label')
```

sphuber commented on docs/source/howto/data.rst:
```diff
@@ -805,3 +805,66 @@ This command will delete both the file repository and the database.
+        ... transfer_builder.source_nodes = {'node_keyname': source_node}
+        ... transfer_builder.metadata.computer = computer_node
+
+The variables ``source_node`` and ``computer_node`` must already contain the corresponding nodes.
```
```suggestion
The variable ``source_node`` here corresponds to the ``RemoteData`` node whose contents need to be retrieved.
```

sphuber commented on docs/source/howto/data.rst:
```diff
@@ -805,3 +805,66 @@ This command will delete both the file repository and the database.
+.. code-block:: ipython
+
+    In [2]: from aiida.engine import submit
+        ... submit( transfer_builder )
```
```suggestion
        ... submit(transfer_builder)
```

sphuber commented on docs/source/howto/data.rst:
```diff
@@ -805,3 +805,66 @@ This command will delete both the file repository and the database.
+        ... instructions_cont['local_files'] = [
+        ...     ('node_keyname', 'source/path/filename', 'target/path/filename'),
+        ... ]
+        ... instructions_node = orm.Dict( dict = instructions_cont )
```
```suggestion
        ... instructions_node = orm.Dict(dict=instructions_cont)
```

sphuber commented on docs/source/howto/data.rst:
```diff
@@ -805,3 +805,66 @@ This command will delete both the file repository and the database.
+        ... ]
+        ... instructions_node = orm.Dict( dict = instructions_cont )
+
+In both cases, you can copy many files by appending them to the respective list.
```
```suggestion
Both when uploading or retrieving, you can copy multiple files by appending them to the list of the ``local_files`` or ``symlink_files`` keys in the instructions input, respectively.
```

sphuber reviewed (changes requested):
Thanks @ramirezfranciscof . Since the docs are new, I still have to ask some changes I'm afraid. I also noted that the input validators are not tested at all for incorrect output. Would be important to quickly add this. Should be straightforward since they are loose functions.

ramirezfranciscof commented on aiida/calculations/transfer.py:
```diff
@@ -0,0 +1,254 @@
+from aiida.common.datastructures import CalcInfo
+
+
+def validate_instructions(instructions, _):
```
So, since one of the checks sees that the computer provided directly is the same as the ones of the `RemoteData` nodes, I need to set up a new computer and I was trying to look for how it is done in the `aiida_localhost` fixture but I can't find it anywhere by grepping. Where is this defined??

sphuber replied:
https://github.com/aiidateam/aiida-core/blob/a873332b80914658c45f588c7d2ce7be79053289/aiida/manage/tests/pytest_fixtures.py#L95

ramirezfranciscof replied:
Uh, I see, thanks. Question: why is this in `aiida/manage/tests` instead of `tests/...`?

sphuber replied:
Because in this way it is also importable by plugins to use directly in their testing frameworks.
https://github.com/aiidateam/aiida-quantumespresso/blob/2c6d02f69cf273cb5f78f6f9f39839d6397793ae/tests/conftest.py#L11

ramirezfranciscof commented:
@sphuber cancelled tests?? wot?

sphuber commented:
> @sphuber cancelled tests?? wot?

It got killed because it ran longer than the 30 min timelimit. Looks like it is a random one off. I restarted them
